### ISSUE TITLE
Poodll recording javascript errors

### ISSUE THREAD

**Issue by ewout** [justinhunt/moodle-filter_poodll#26]
After the update to poodll 3.0.38(Build 2017041703) the html5 recording stopped working.

As before, recording doesn't work in Firefox (Ubuntu), but now it also stopped working in Chrome (57) Ubuntu. In the console I get 
```
first.js:82 media error 
Object {name: "TrackStartError", message: "", constraint: "", toString: function}
onMediaError	@	first.js:82
```

**Comment by justinhunt** [comment]
OK. There was quite a significant change in this area in 2017041702 
https://github.com/justinhunt/moodle-filter_poodll/commit/d0b3375083b0d0998dda4454a0ae8705c210715a
Its probably related to that. I actually tested both Firefox and Chrome on Xubuntu. But this area is a bit rocky and success depends on device availability.

If you did not already, can you try to "purge cache" from your Moodle site's admin area. Even after the upgrade this does not always happen, and it will cause issues in this case.

If that doesn't make it better, please post your support info file  (see the poodll filter settings) at https://poodll.com/contact and I can see the specifics

**Comment by ewout** [comment]
I can confirm that on a computer without a webcam, the same type of errors (the popup happens, but the buttons don't work and erros in the webdev console) occurred. But later it started working again with a webcam. Now on a laptop with good webcam, again problems. 

It all seems very fragile.

**Comment by justinhunt** [comment]
And have you purged the cache?

**Comment by ewout** [comment]
Yeah, man. First thing I did :). But now I purged it again, and low an behold, it works!

Could it be I dreamed I purged the cache...? I really don't think so. 

I'll do more testing this weekend, if you want to close this ticket, that's fine by me.

**Comment by justinhunt** [comment]
Lets leave it on open for a bit. Fragile looks like this ...

**Comment by nadavkav** [comment]
I am also getting this (clicking "record") on a poodll video recording question type, in a quiz.

I am using latest Chrome Version 57.0.2987.133 (64-bit) on a Kubuntu GNU/Linux OS).
Moodle 3.3 + Poodll 2017082601

**Comment by justinhunt** [comment]
Nadav. This is a client side error. I  think its probably to do with access to the mic/webcam being unavailable. Due to another app using it, or a driver issue or driver crash.  A browser reboot, or system reboot in that case, would hopefully "fix" it.  Can you see if thats the case? We need to put better error reporting in for this sort of thing

**Comment by nadavkav** [comment]
I think you are right :( as it happens on a Chrome v57 browser, but works fine on Firefox v57.
Restarting the browser does not help.

You might be interested to know that: I have installed https://moodle.org/plugins/atto_recordrtc which works fine on both browsers.

BTW, I got this complaint from our teachers and students, that uses different Chrome versions on a Windows machines. (not sure what Windows versions and what Chrome versions)

**Comment by nadavkav** [comment]
Thank you for your quick reply :smile:

**Comment by justinhunt** [comment]
Nadav. Does it work if you use the (almost) latest Poodll at https://demo.poodll.com ? We added a small settings icon on the lower right of each recorder, allowing you to alter the selected audio or video device. I can not recreate the issue so its a bit hard to test, but perhaps you could see if 
a) the new code is more reliable
b) if toggling the selected video device helps

**Comment by justinhunt** [comment]
Also ... this could be related to the audio stream more so than the video stream. Because we assume video requires audio so request both tracks. But when the audio track fails, because there is a cam and no mic, I think this error occurs. Note, that 'no mic' can mean no mic is available even though there is one on the system.

**Comment by nadavkav** [comment]
I tested it on the quiz inside demo.poodll.com, and I am getting the same error. ( It recognizes the video device)
Also, noticed there is no audio device I can choose to use. and that might be the issue.
I will do more digging and report back here, soon.
